### PR TITLE
SONARJAVA-2482 New rule S4266 RedundantStreamCollectCheck

### PR DESCRIPTION
**Pull request by saberduck**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/RedundantStreamCollectCheck.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantStreamCollectCheck.java]: Code Smell Code Smell: Remove this unused import 'java.util.Optional'. ([squid:UselessImportCheck](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.java%3Ajava&branch=SONARJAVA-2482&issues=AWDmNvGeGaMZFZs6hiio&open=AWDmNvGeGaMZFZs6hiio)

[ISSUE_KEY]: AWDmNvGeGaMZFZs6hiio

File: java-checks/src/main/java/org/sonar/java/checks/RedundantStreamCollectCheck.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.commons.lang.reflect.MethodUtils;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.helpers.MethodsHelper;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.java.matcher.MethodMatcherCollection;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantStreamCollectCheck.java]: Code Smell Code Smell: Remove this unused import 'org.sonar.plugins.java.api.tree.MemberSelectExpressionTree'. ([squid:UselessImportCheck](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.java%3Ajava&branch=SONARJAVA-2482&issues=AWDmNvGeGaMZFZs6hiim&open=AWDmNvGeGaMZFZs6hiim)

[ISSUE_KEY]: AWDmNvGeGaMZFZs6hiim

File: java-checks/src/main/java/org/sonar/java/checks/RedundantStreamCollectCheck.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.commons.lang.reflect.MethodUtils;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.helpers.MethodsHelper;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.java.matcher.MethodMatcherCollection;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantStreamCollectCheck.java]: Code Smell Code Smell: Remove this unused import 'org.sonar.java.matcher.MethodMatcherCollection'. ([squid:UselessImportCheck](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.java%3Ajava&branch=SONARJAVA-2482&issues=AWDmNvGeGaMZFZs6hiip&open=AWDmNvGeGaMZFZs6hiip)

[ISSUE_KEY]: AWDmNvGeGaMZFZs6hiip

File: java-checks/src/main/java/org/sonar/java/checks/RedundantStreamCollectCheck.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.commons.lang.reflect.MethodUtils;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantStreamCollectCheck.java]: Code Smell Code Smell: Remove this unused import 'org.apache.commons.lang.reflect.MethodUtils'. ([squid:UselessImportCheck](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.java%3Ajava&branch=SONARJAVA-2482&issues=AWDmNvGeGaMZFZs6hiin&open=AWDmNvGeGaMZFZs6hiin)

[ISSUE_KEY]: AWDmNvGeGaMZFZs6hiin

File: java-checks/src/main/java/org/sonar/java/checks/RedundantStreamCollectCheck.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  private static final String COLLECTORS = "java.util.stream.Collectors";
+
+  private static final MethodMatcher COUNTING = MethodMatcher.create().typeDefinition(COLLECTORS).name("counting").withoutParameter();
+  private static final MethodMatcher MAX_BY = MethodMatcher.create().typeDefinition(COLLECTORS).name("maxBy").withAnyParameters();
+  private static final MethodMatcher MIN_BY = MethodMatcher.create().typeDefinition(COLLECTORS).name("minBy").withAnyParameters();
+  private static final MethodMatcher MAPPING = MethodMatcher.create().typeDefinition(COLLECTORS).name("mapping").withAnyParameters();
+  private static final MethodMatcher REDUCING = MethodMatcher.create().typeDefinition(COLLECTORS).name("reducing").withAnyParameters();
+  private static final MethodMatcher SUMMING_INT = MethodMatcher.create().typeDefinition(COLLECTORS).name("summingInt").withAnyParameters();
+  private static final MethodMatcher SUMMING_LONG = MethodMatcher.create().typeDefinition(COLLECTORS).name("summingLong").withAnyParameters();
+  private static final MethodMatcher SUMMING_DOUBLE = MethodMatcher.create().typeDefinition(COLLECTORS).name("summingDouble").withAnyParameters();
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantStreamCollectCheck.java]: I would maybe extract `MethodMatcher.create().typeDefinition(COLLECTORS).name(...)` into a dedicated static method such as:
```
private static collectorsMethod(String name) {
  return MethodMatcher.create().typeDefinition(COLLECTORS).name(name);
}

to reduce a bit all the repeated code.

**Pull request merged**